flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java

```diff
@@ -590,6 +604,15 @@ public Execution resetForNewExecution(final long timestamp, final long originati
 
 				this.currentExecution = newExecution;
 
+				synchronized (this.inputSplits){
+					for (InputSplit split: this.inputSplits){
+						if (split != null) {
+							this.jobVertex.getSplitAssigner().returnInputSplit(split, this.getParallelSubtaskIndex());
```

Comment:
From the usage, it seems like `returnInputSplit` should better be a bulk operation that takes a `Collection<InputSplit>`.

Reply:
Done.